Add the json-deserializer middleware package

matt-jenner opened the pull request:
This implements a json-deserializer middleware, it expects an object extending ApiGatewayProxyEvent and returns an ApiGatewayProxyObjectEvent which replaces the string body for an object of type `Record<string, unknown>`.

It introduces the concept of a `RequestBodyNotJsonError` to be used to catch and return an appropriate 400 error back if a non json payload is passed.

If no body or a non-json header is passed it will pass null as the body property on the event passed to the underlying handler.

Closes #54 

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

I'm sorry for the delays in supplying this PR, but things got crazily busy at work, and then summer (and a large DIY project - building a garden office now my work has decided no-one is going back to the office full-time) happened which I'm afraid took most of my spare time and I like to code when fully able to commit myself to it.

Although I mostly got this done within a few hours work.  I ran into an issue with the 100% code-coverage requirements which caused me to delay submiting.  While I do have full coverage in TS (and I always like to see with libraries I expect someone to depend upon), there was one issue where the `super(message)` line in the `RequestBodyNotJsonError` class would NOT mark itself as covered despite definately hitting the line in my tests.

Jest reports the line as not-covered due to the coverage being applied to ES5 code rather than the TS itself.  I've confirmed coverage is 100% as if you switch the output to ES6 it reports 100% coverage.

After reading quite a bit around the issue, it seems to relate to the fact that when the TS line:
`super(message)`

 is transpiled down to ES5, it comes out as:
`var _this = _super.call(this, message) || this;`

While the former part of the or statement is covered, it doesn't cover all the cases of the transpiled statement fully.

I've tried various approaches to resolve this including a suggestion of adding `https://www.npmjs.com/package/ts-es5-istanbul-coverage` but eventually thought it best to submit it cleanly without this and see your thoughts.

See here for more information:
- https://github.com/microsoft/TypeScript/issues/13029
- https://github.com/SitePen/remap-istanbul/issues/106

Otherwise, hopefully the PR should be otherwise mostly self-explanatory, the logic itself is very simple and is the precursor for my other feature request logged in #55.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dbartholomae/lambda-middleware?pullRequest=57) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dbartholomae/lambda-middleware?pullRequest=57) before we can accept your contribution.<br/><hr/>**Matt Jenner** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dbartholomae/lambda-middleware?pullRequest=57) it.</sub>

dbartholomae commented on packages/json-deserializer/src/JsonDeserializer.ts:
```diff
@@ -0,0 +1,57 @@
+) => async (event: E, context: Context): Promise<APIGatewayProxyResult> => {
+  let body;
+
+  body = deserializeBody(event);
```
Could this just be
```suggestion
  const body = deserializeBody(event);
```

matt-jenner replied:
Hmm yes, I'm not sure why this isn't a const!? I'll fix it shortly.

dbartholomae reviewed (commented):
Won't be able to do a full review until I'm back from vacation, but here are already some observations

dbartholomae commented:
Concerning the coverage, just adding a comment explaining it and an Istanbul ignore next might well suffice

matt-jenner commented on packages/json-deserializer/src/JsonDeserializer.test.ts:
```diff
@@ -0,0 +1,98 @@
+    },
+  };
+
+  it("should deserialize objects if the event has a json content-type header", async () => {
```
Sure, in the past I've had people ask for all the tests to contain should in the past on other projects to reflect expected behaviour, rather than a statement of fact, happy to fit with convention though.

matt-jenner commented on packages/json-deserializer/src/JsonDeserializer.test.ts:
```diff
@@ -0,0 +1,98 @@
+    it.each([
+      ["null headers", null],
+      ["null content type", { "Content-Type": null }],
+      ["a non-json content-type header", { "Content-Type": "text/plain" }],
```
Hmm, good thought, I think by using more opinionated middlewares that do more work to the payload or other aspects of the event, you are narrowing down to a more opinionated set of behaviours.  

In this case though, we could potentially make it emit an object with an additional `bodyObject` property rather than replacing the body, and that way in the case of a non json content header it could just pass through the payload.  Replacing the body just felt cleaner but I think maybe it might make it more flexible if I move the deserialized body out to a separate property? 

In this case of these tests, it would simply pass through without failing, but set a null `bodyObject` property.

If you're in agreement I can make these modifications.

matt-jenner replied:
I've decided to switch to `bodyObject` this now makes it non-destructive if added to a separate middleware that still needs the body property.

matt-jenner commented on packages/json-deserializer/src/JsonDeserializer.test.ts:
```diff
@@ -0,0 +1,98 @@
+    expect(processedEvent.body).toBeNull();
+  });
+
+  it("should throw a RequestBodyNotJsonError if the request body cannot deserialize", async () => {
```
I think at the time I was thinking "what about GET scenarios" if I were to setup my middlewares generically outside of the handlers and just pass the handler to a factory function that wraps a handler in standard middlewares for the project if you'd like to standardise them rather than repeating adding all the layers in handlers where it's needed.  There's no harm as such in returning null in this case.

Thinking about this, I think maybe adding a separate `bodyObject` rather than replacing `body` might be better with this behaviour as the null functionality is halfway towards making it more "compatible" with varying usage.

matt-jenner replied:
I've not changed the behaviour for the reasons above, but I've implemented the `bodyObject` instead of overwriting the body, that will probably make the behaviour make more sense.

matt-jenner commented on packages/json-deserializer/src/JsonDeserializer.test.ts:
```diff
@@ -0,0 +1,98 @@
+
+    expect(() =>
+      handlerWithMiddleware(event, createContext())
+    ).rejects.toMatchObject({
```
Currently, no it does not contain a statusCode property as I have tried to keep errors themselves separate from http response logic in past projects but seeing as we're emitting this error from within the middleware, I think should be fine to do that.  I've taken a look at the error handler just now, it should be simple enough to add a 400 statusCode into the error.

dbartholomae replied:
```suggestion
    ).rejects.toMatchObject({
      statusCode: 400,
```

dbartholomae commented on packages/json-deserializer/src/customErrors/RequestBodyNotJsonError.ts:
```diff
@@ -0,0 +1,15 @@
+export class RequestBodyNotJsonError extends Error {
+  body: string;
+  innerError: Error;
```
```suggestion
  innerError: Error;
  statusCode: 400;
```

dbartholomae commented on packages/json-deserializer/src/customErrors/RequestBodyNotJsonError.ts:
```diff
@@ -0,0 +1,15 @@
+
+    this.name = "RequestBodyNotJsonError";
+    this.body = body;
+    this.innerError = innerError;
```
```suggestion
    this.innerError = innerError;
    this.statusCode = 400;
```

dbartholomae reviewed (changes requested):
Thanks for waiting! I've added comments to make it easy to add the status code to the error. I'll also check why CI isn't running on this PR. There might need to run `rush change` to also add a changelog. But CI should tell us.

dbartholomae commented:
Please also rebase, this should run CI for the branch (I've just updated the main branch with a small config change).

matt-jenner commented:
Thanks for the review, I hope you had a good holiday.  I've added your suggestions and fixed one codeQL vulnerability relating to Regular expression denial of service (it was simpler to just remove the regex 🙂).

It's running the build now but for some reason it's not seeing my package in the build, it lists 13 packages not the 14 I see if I run `rush test:integration` or `rush test:unit` locally and the project is listed in `rush.json` so I would have thought it would have picked it up?

Any ideas on this front as I'm afraid I'm not a rush expert, thanks 🙂.

matt-jenner commented:
Thanks for the approval - it will feel good to finally get this in, I think you need to merge this manually though as mergify is complaining that the workflow has changed.  Once this is in, I can start on the next issue I logged to create the io-ts validator middleware before I'm due back in work next week, thanks 🙂.

dbartholomae commented:
Thanks! I now figured out the problem: When I changed the settings to run the GitHub actions for this PR, I changed them in a way that lead to actions running on the main branch code instead of the PR branch code instead. I've just updated the main branch to fix this. If you could rebase/merge main one more time, it should be solved :)

dbartholomae commented:
Just realized that you gave me access to the branch, so I did the merge of main myself and will merge this branch once CI is green.

matt-jenner commented:
> 
> 
> There was also a slight problem with the integration test: it still used `body` instead of `bodyObject`. It's fixed, and version 1.1.0 is now available on npm.

Thanks for fixing that, I really mustn't code quite so late 🙂 looking forward to switching my project over to use this now and start on the next middleware 😁
